File: 032.rb
# 032
# Longest Valid ParenthesesXXX
# 2015-05-10
#####################################################
# @param {String} s
# @return {Integer}
def longest_valid_parentheses(s)
  stack = [-1]
  count = 0
  for i in (0...s.size)
    if s[i] == '('
      stack.push(i)
    elsif stack.last != -1 and s[stack.last] == '('
      stack.pop
      count = [count, i - stack.last].max
    else
      stack.push(i)
    end
  end
  count
end

#####################################################

s = "()()"
puts longest_valid_parentheses(s)

File: 032.py
#   032
#   Longest Valid ParenthesesXXX
#   2015-05-10
#####################################################
class Solution:
    # @param {string} s
    # @return {integer}
    def longestValidParentheses(self, s):
        count = 0
        stack = [-1]
        for i in range(0, len(s)):
            if s[i] == '(':
                stack.append(i)
            elif stack[-1] != -1 and s[stack[-1]] == '(':
                stack.pop()
                count = max(count, i-stack[-1])
            else:
                stack.append(i)
        return count

#####################################################

s = "()()"
print s

sol = Solution()
print sol.longestValidParentheses(s)

